runtime/jcl/common/jclcinit.c

```diff
@@ -712,6 +716,58 @@ initializeRequiredClasses(J9VMThread *vmThread, char* dllName)
 		}
 	}
 
+	if ((J9SYSPROP_ERROR_NONE == vm->internalVMFunctions->getSystemProperty(vm, ADD_MGMT_AGENT_MODULE_PROP, NULL))
+			&& (NULL != vm->modulesPathEntry->extraInfo)) {
+		Trc_JCL_initializeRequiredClasses_addAgentModuleEntry(vmThread);
+		const char *module = vm->jimageIntf->jimagePackageToModule(
+				vm->jimageIntf, (UDATA) vm->modulesPathEntry->extraInfo,
+				"jdk/internal/agent");
+		if (NULL != module) {
+			J9VMSystemProperty *systemProperty = NULL;
+			UDATA propIndex = 0;
+			BOOLEAN done = FALSE;
+			/* Handle the case where there are no user-specified modules. In this case, there
+			 * is typically one system property set to "ALL-DEFAULT".
+			 */
+			if (J9SYSPROP_ERROR_NONE == vm->internalVMFunctions->getSystemProperty(vm, ADDMODS_PROPERTY_BASE "0", &systemProperty)) {
+				if (0 == strcmp(systemProperty->value, "ALL-DEFAULT")) {
+					/* this is implicitly an unused property */
+					vm->internalVMFunctions->setSystemProperty(vm, systemProperty, AGENT_MODULE_NAME);
+					done = TRUE;
+				} else {
+					propIndex = 1;
+				}
+			}
+			if (!done) {
+				char *propIndexPosition = NULL;
+				/* Allow up to 10000 --add-modules arguments. */
+				char *propNameBuffer = j9mem_allocate_memory(sizeof(ADDMODS_PROPERTY_BASE) + 4, OMRMEM_CATEGORY_VM);
+
+				if (NULL == propNameBuffer) {
+					Trc_JCL_initializeRequiredClasses_addAgentModuleOutOfMemory(vmThread);
+					return 1;
+				}
+				propIndexPosition = propNameBuffer + strlen(ADDMODS_PROPERTY_BASE);
+				j9str_printf(PORTLIB, propNameBuffer, sizeof(ADDMODS_PROPERTY_BASE), ADDMODS_PROPERTY_BASE);
+				while ((propIndex < 10000) && !done) {
```

Comment:
The code which adds the options is here
https://github.com/eclipse/openj9/blob/bf3df3839011c12267f31ae1006f8bb58a408f8e/runtime/vm/vmprops.c#L431

You can modify addPropertiesForOptionWithAssignArg() to take a parameter that returns the index, and save the index in the J9JavaVM.